File: Part7/Go/Maze_N_x_N/Maze_N_x_N.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
    "time"
)

func main ( ) {
    const FIRST_CELL = 0
    const SECOND_CELL = 1
    const NO_CELL = -1
    
    scanner := bufio.NewScanner ( os.Stdin )
    
    // Prompt the user for maze size
    var mazeColumns int
    for ( mazeColumns <= 0 ) {
        fmt.Print ( "Please enter number of columns for maze, must be greater than 1: " )
        scanner.Scan ( )
        userInput := strings.TrimSpace ( scanner.Text ( ) )
        
        if num, err := strconv.Atoi ( userInput ); err == nil {
            mazeColumns = num
        }
    }
    
    var mazeRows int
    for ( mazeRows <= 0 ) {
        fmt.Print ( "Please enter number of rows for maze, must be greater than 1: " )
        scanner.Scan ( )
        userInput := strings.TrimSpace ( scanner.Text ( ) )
        
        if num, err := strconv.Atoi ( userInput ); err == nil {
            mazeRows = num
        }
    }
    
    interiorWallCount := mazeRows * ( mazeColumns - 1 ) + ( mazeRows - 1 ) * mazeColumns
    
    // Start with all the walls up
    wallsUp := make ( []bool, interiorWallCount )
    for i := range wallsUp {
        wallsUp[i] = true
    }
    
    // Identify the cells each wall connects
    wallConnections := make ( [][2]int, interiorWallCount )
    
    wallIndex := 0
    for rowIndex := 0; rowIndex < mazeRows; rowIndex++ {
        firstCellInRow := rowIndex * mazeColumns
        
        // Vertical walls
        for verticalWallIndex := 0; verticalWallIndex < mazeColumns - 1; verticalWallIndex++ {
            leftCell := firstCellInRow + verticalWallIndex
            rightCell := leftCell + 1
            wallConnections[wallIndex][FIRST_CELL] = leftCell
            wallConnections[wallIndex][SECOND_CELL] = rightCell
            wallIndex++
        }
        
        // Horizontal walls
        if ( wallIndex < interiorWallCount ) {
            for horizontalWallIndex := 0; horizontalWallIndex < mazeColumns; horizontalWallIndex++ {
                upperCell := firstCellInRow + horizontalWallIndex
                lowerCell := upperCell + mazeColumns
                wallConnections[wallIndex][FIRST_CELL] = upperCell
                wallConnections[wallIndex][SECOND_CELL] = lowerCell
                wallIndex++
            }
        }
    }
    
    // Identify which group each cell is a part of
    cellToGroup := make ( []int, mazeRows * mazeColumns )
    for i := range cellToGroup {
        cellToGroup[i] = i
    }
    
    // Identify which cells are a part of each group
    groupCells := make ( [][]int, mazeColumns * mazeRows )
    for i := range groupCells {
        groupCells[i] = make ( []int, mazeColumns * mazeRows )
        for j := range groupCells[i] {
            if ( j == 0 ) {
                groupCells[i][j] = i
            } else {
                groupCells[i][j] = NO_CELL
            }
        }
    }
    
    // Print initial maze
    currentInteriorWall := 0
    
    // Print top border
    fmt.Print ( "+" )
    for i := 0; i < mazeColumns; i++ {
        fmt.Print ( "-+" )
    }
    fmt.Println ( )
    
    for rowIndex := 0; rowIndex < mazeRows; rowIndex++ {
        // Print vertical walls and cells
        fmt.Print ( "|" )
        for columnIndex := 0; columnIndex < mazeColumns; columnIndex++ {
            fmt.Print ( " " )
            
            if ( columnIndex == mazeColumns - 1 || wallsUp[currentInteriorWall] ) {
                fmt.Print ( "|" )
            } else {
                fmt.Print ( " " )
            }
            
            if ( columnIndex < mazeColumns - 1 ) {
                currentInteriorWall++
            }
        }
        fmt.Println ( )
        
        // Print horizontal walls
        if ( rowIndex < mazeRows - 1 ) {
            fmt.Print ( "+" )
            for columnIndex := 0; columnIndex < mazeColumns; columnIndex++ {
                if ( wallsUp[currentInteriorWall] ) {
                    fmt.Print ( "-" )
                } else {
                    fmt.Print ( " " )
                }
                fmt.Print ( "+" )
                currentInteriorWall++
            }
            fmt.Println ( )
        }
    }
    
    // Print bottom border
    fmt.Print ( "+" )
    for i := 0; i < mazeColumns; i++ {
        fmt.Print ( "-+" )
    }
    fmt.Println ( )
    
    // Create wall removal list
    wallRemoveList := make ( []int, interiorWallCount )
    for i := 0; i < interiorWallCount; i++ {
        wallRemoveList[i] = i
    }
    
    // Seed the random number generator
    seed := time.Now ( ).UnixNano ( )
    
    // Fisher-Yates shuffle algorithm
    for shuffleIndex := interiorWallCount - 1; shuffleIndex > 0; shuffleIndex-- {
        // Simple LCG for random number generation
        seed = seed * 1664525 + 1013904223
        otherIndex := int ( seed ) % ( shuffleIndex + 1 )
        if ( otherIndex < 0 ) {
            otherIndex = -otherIndex
        }
        
        temp := wallRemoveList[shuffleIndex]
        wallRemoveList[shuffleIndex] = wallRemoveList[otherIndex]
        wallRemoveList[otherIndex] = temp
    }
    
    mazeComplete := false
    
    // Kruskal's algorithm
    for removeWallIndex := 0; removeWallIndex < interiorWallCount; removeWallIndex++ {
        nextWallToCheck := wallRemoveList[removeWallIndex]
        
        firstCell := wallConnections[nextWallToCheck][FIRST_CELL]
        firstCellGroupIndex := cellToGroup[firstCell]
        secondCell := wallConnections[nextWallToCheck][SECOND_CELL]
        secondCellGroupIndex := cellToGroup[secondCell]
        
        if ( firstCellGroupIndex != secondCellGroupIndex ) {
            wallsUp[nextWallToCheck] = false
            
            // Find next empty position in first group
            nextEmptyFirstGroupIndex := 0
            for cellIndex := 0; cellIndex < mazeColumns * mazeRows; cellIndex++ {
                if ( groupCells[firstCellGroupIndex][cellIndex] == NO_CELL ) {
                    nextEmptyFirstGroupIndex = cellIndex
                    break
                }
            }
            
            // Move all cells from second group to first group
            for groupCellIndex := mazeColumns * mazeRows - 1; groupCellIndex >= 0; groupCellIndex-- {
                if ( groupCells[secondCellGroupIndex][groupCellIndex] != NO_CELL ) {
                    cellToMove := groupCells[secondCellGroupIndex][groupCellIndex]
                    
                    groupCells[firstCellGroupIndex][nextEmptyFirstGroupIndex] = cellToMove
                    nextEmptyFirstGroupIndex++
                    cellToGroup[cellToMove] = firstCellGroupIndex
                    groupCells[secondCellGroupIndex][groupCellIndex] = NO_CELL
                    
                    if ( nextEmptyFirstGroupIndex >= mazeColumns * mazeRows ) {
                        mazeComplete = true
                    }
                }
            }
            
            time.Sleep ( 500 * time.Millisecond )
            
            // Copy in Print maze code from above again here:
            currentInteriorWall = 0
            
            fmt.Println ( )
            fmt.Print ( "+" )
            for i := 0; i < mazeColumns; i++ {
                fmt.Print ( "-+" )
            }
            fmt.Println ( )
            
            for rowIndex := 0; rowIndex < mazeRows; rowIndex++ {
                fmt.Print ( "|" )
                for columnIndex := 0; columnIndex < mazeColumns; columnIndex++ {
                    fmt.Print ( " " )
                    
                    if ( columnIndex == mazeColumns - 1 || wallsUp[currentInteriorWall] ) {
                        fmt.Print ( "|" )
                    } else {
                        fmt.Print ( " " )
                    }
                    
                    if ( columnIndex < mazeColumns - 1 ) {
                        currentInteriorWall++
                    }
                }
                fmt.Println ( )
                
                if ( rowIndex < mazeRows - 1 ) {
                    fmt.Print ( "+" )
                    for columnIndex := 0; columnIndex < mazeColumns; columnIndex++ {
                        if ( wallsUp[currentInteriorWall] ) {
                            fmt.Print ( "-" )
                        } else {
                            fmt.Print ( " " )
                        }
                        fmt.Print ( "+" )
                        currentInteriorWall++
                    }
                    fmt.Println ( )
                }
            }
            
            fmt.Print ( "+" )
            for i := 0; i < mazeColumns; i++ {
                fmt.Print ( "-+" )
            }
            fmt.Println ( )
            
            if ( mazeComplete ) {
                break
            }
        }
    }
}

File: Part7/Python/Maze_N_x_N/Maze_N_x_N.py
import random
import time

FIRST_CELL = 0
SECOND_CELL = 1
NO_CELL = -1

# Prompt the user for maze size
mazeColumns = 0
while ( mazeColumns <= 0 ):
    userInput = input ( "Please enter number of columns for maze, must be greater than 1: " )
    # Check if string contains only digits
    isValid = True
    if ( len ( userInput ) == 0 ):
        isValid = False
    else:
        i = 0
        while ( i < len ( userInput ) ):
            if ( userInput[i] < '0' or userInput[i] > '9' ):
                isValid = False
                break
            i = i + 1
    
    if ( isValid ):
        mazeColumns = int ( userInput )

mazeRows = 0
while ( mazeRows <= 0 ):
    userInput = input ( "Please enter number of rows for maze, must be greater than 1: " )
    # Check if string contains only digits
    isValid = True
    if ( len ( userInput ) == 0 ):
        isValid = False
    else:
        i = 0
        while ( i < len ( userInput ) ):
            if ( userInput[i] < '0' or userInput[i] > '9' ):
                isValid = False
                break
            i = i + 1
    
    if ( isValid ):
        mazeRows = int ( userInput )

interiorWallCount = mazeRows * ( mazeColumns - 1 ) + ( mazeRows - 1 ) * mazeColumns

# Start with all the walls up
wallsUp = [False] * interiorWallCount
i = 0
while ( i < interiorWallCount ):
    wallsUp[i] = True
    i = i + 1

# Identify the cells each wall connects
wallConnections = [[0, 0] for _ in range ( interiorWallCount )]

wallIndex = 0
rowIndex = 0
while ( rowIndex < mazeRows ):
    # Track the first cell in the current row
    firstCellInRow = rowIndex * mazeColumns
    
    # Vertical walls
    verticalWallIndex = 0
    while ( verticalWallIndex < mazeColumns - 1 ):
        leftCell = firstCellInRow + verticalWallIndex
        rightCell = leftCell + 1
        wallConnections[wallIndex][FIRST_CELL] = leftCell
        wallConnections[wallIndex][SECOND_CELL] = rightCell
        wallIndex = wallIndex + 1
        verticalWallIndex = verticalWallIndex + 1
    
    # Horizontal walls
    if ( wallIndex < interiorWallCount ):
        horizontalWallIndex = 0
        while ( horizontalWallIndex < mazeColumns ):
            upperCell = firstCellInRow + horizontalWallIndex
            lowerCell = upperCell + mazeColumns
            wallConnections[wallIndex][FIRST_CELL] = upperCell
            wallConnections[wallIndex][SECOND_CELL] = lowerCell
            wallIndex = wallIndex + 1
            horizontalWallIndex = horizontalWallIndex + 1
    
    rowIndex = rowIndex + 1

# Identify which group each cell is a part of
cellToGroup = [0] * ( mazeRows * mazeColumns )
cellIndex = 0
while ( cellIndex < mazeRows * mazeColumns ):
    cellToGroup[cellIndex] = cellIndex
    cellIndex = cellIndex + 1

# Identify which cells are a part of each group
groupCells = [[NO_CELL for _ in range ( mazeRows * mazeColumns )] for _ in range ( mazeRows * mazeColumns )]

cellIndex = 0
while ( cellIndex < mazeRows * mazeColumns ):
    innerCellIndex = 0
    while ( innerCellIndex < mazeRows * mazeColumns ):
        if ( innerCellIndex == 0 ):
            groupCells[cellIndex][innerCellIndex] = cellIndex
        else:
            groupCells[cellIndex][innerCellIndex] = NO_CELL
        innerCellIndex = innerCellIndex + 1
    cellIndex = cellIndex + 1

# Print initial maze
currentInteriorWall = 0

# Print top border
print ( "+", end="" )
cellIndex = 0
while ( cellIndex < mazeColumns ):
    print ( "-+", end="" )
    cellIndex = cellIndex + 1
print ( )

rowIndex = 0
while ( rowIndex < mazeRows ):
    # Print vertical walls and cells
    print ( "|", end="" )
    columnIndex = 0
    while ( columnIndex < mazeColumns ):
        print ( " ", end="" )
        
        if ( columnIndex == mazeColumns - 1 or wallsUp[currentInteriorWall] ):
            print ( "|", end="" )
        else:
            print ( " ", end="" )
        
        if ( columnIndex < mazeColumns - 1 ):
            currentInteriorWall = currentInteriorWall + 1
        columnIndex = columnIndex + 1
    print ( )
    
    # Print horizontal walls
    if ( rowIndex < mazeRows - 1 ):
        print ( "+", end="" )
        columnIndex = 0
        while ( columnIndex < mazeColumns ):
            if ( wallsUp[currentInteriorWall] ):
                print ( "-", end="" )
            else:
                print ( " ", end="" )
            print ( "+", end="" )
            currentInteriorWall = currentInteriorWall + 1
            columnIndex = columnIndex + 1
        print ( )
    rowIndex = rowIndex + 1

# Print bottom border
print ( "+", end="" )
columnIndex = 0
while ( columnIndex < mazeColumns ):
    print ( "-+", end="" )
    columnIndex = columnIndex + 1
print ( )

# Create wall removal list
wallRemoveList = [0] * interiorWallCount
wallIndex = 0
while ( wallIndex < interiorWallCount ):
    wallRemoveList[wallIndex] = wallIndex
    wallIndex = wallIndex + 1

# Fisher-Yates shuffle algorithm
shuffleIndex = interiorWallCount - 1
while ( shuffleIndex > 0 ):
    otherIndex = random.randint ( 0, shuffleIndex )
    temp = wallRemoveList[shuffleIndex]
    wallRemoveList[shuffleIndex] = wallRemoveList[otherIndex]
    wallRemoveList[otherIndex] = temp
    shuffleIndex = shuffleIndex - 1

mazeComplete = False

# Kruskal's algorithm
removeWallIndex = 0
while ( removeWallIndex < interiorWallCount ):
    nextWallToCheck = wallRemoveList[removeWallIndex]
    
    firstCellValue = wallConnections[nextWallToCheck][FIRST_CELL]
    firstCellGroupIndex = cellToGroup[firstCellValue]
    secondCellValue = wallConnections[nextWallToCheck][SECOND_CELL]
    secondCellGroupIndex = cellToGroup[secondCellValue]
    
    if ( firstCellGroupIndex != secondCellGroupIndex ):
        wallsUp[nextWallToCheck] = False
        
        # Find next empty position in first group
        nextEmptyFirstGroupIndex = 0
        cellIndex = 0
        while ( cellIndex < mazeColumns * mazeRows ):
            if ( groupCells[firstCellGroupIndex][cellIndex] == NO_CELL ):
                nextEmptyFirstGroupIndex = cellIndex
                break
            cellIndex = cellIndex + 1
        
        # Move all cells from second group to first group
        groupCellIndex = mazeColumns * mazeRows - 1
        while ( groupCellIndex >= 0 ):
            if ( groupCells[secondCellGroupIndex][groupCellIndex] != NO_CELL ):
                cellToMove = groupCells[secondCellGroupIndex][groupCellIndex]
                
                groupCells[firstCellGroupIndex][nextEmptyFirstGroupIndex] = cellToMove
                nextEmptyFirstGroupIndex = nextEmptyFirstGroupIndex + 1
                cellToGroup[cellToMove] = firstCellGroupIndex
                groupCells[secondCellGroupIndex][groupCellIndex] = NO_CELL
                
                if ( nextEmptyFirstGroupIndex >= mazeColumns * mazeRows ):
                    mazeComplete = True
            groupCellIndex = groupCellIndex - 1
        
        time.sleep ( 0.5 )
        
        # Copy in Print maze code from above again here:
        currentInteriorWall = 0
        
        print ( )
        print ( "+", end="" )
        cellIndex = 0
        while ( cellIndex < mazeColumns ):
            print ( "-+", end="" )
            cellIndex = cellIndex + 1
        print ( )
        
        rowIndex = 0
        while ( rowIndex < mazeRows ):
            print ( "|", end="" )
            columnIndex = 0
            while ( columnIndex < mazeColumns ):
                print ( " ", end="" )
                
                if ( columnIndex == mazeColumns - 1 or wallsUp[currentInteriorWall] ):
                    print ( "|", end="" )
                else:
                    print ( " ", end="" )
                
                if ( columnIndex < mazeColumns - 1 ):
                    currentInteriorWall = currentInteriorWall + 1
                columnIndex = columnIndex + 1
            print ( )
            
            if ( rowIndex < mazeRows - 1 ):
                print ( "+", end="" )
                columnIndex = 0
                while ( columnIndex < mazeColumns ):
                    if ( wallsUp[currentInteriorWall] ):
                        print ( "-", end="" )
                    else:
                        print ( " ", end="" )
                    print ( "+", end="" )
                    currentInteriorWall = currentInteriorWall + 1
                    columnIndex = columnIndex + 1
                print ( )
            rowIndex = rowIndex + 1
        
        print ( "+", end="" )
        columnIndex = 0
        while ( columnIndex < mazeColumns ):
            print ( "-+", end="" )
            columnIndex = columnIndex + 1
        print ( )
        
        if ( mazeComplete ):
            break
    
    removeWallIndex = removeWallIndex + 1
